EIPS/eip-6066.md

```diff
@@ -0,0 +1,127 @@
+---
+eip: 6066
+title: Signature Validation Method for NFTs
+description: A way to verify signatures when the signing entity is an EIP-721 or EIP-1155 NFT
+author: Jack Boyuan Xu (@boyuanx)
+discussions-to: https://ethereum-magicians.org/t/eip-6066-signature-validation-method-for-nfts/
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-11-29
+requires: 721, 1155, 5750
+---
+
+## Abstract
+
+While **E**xternally **O**wned **A**ccounts can validate signed messages with `ecrecover()` and smart contracts can validate signatures using specifications outlined in [EIP-1271](./eip-1271.md), currently there is no standard method to create or validate signatures made by NFTs. We propose a standard way for anyone to validate whether a signature made by an NFT is valid. This is possible via a modified signature validation function originally found in [EIP-1271](./eip-1271.md): `isValidSignature(tokenId, hash, data)`.
+
+## Motivation
+
+With billions of ETH in trading volume, the **N**on-**F**ungible **T**oken standard has exploded into tremendous popularity in recent years. Despite the far-reaching implications of having unique tokenized items on-chain, NFTs have mainly been used to represent artwork in the form of avatars or profile pictures. While this is certainly not a trivial use case for the [EIP-721](./eip-721.md) & [EIP-1155](./eip-1155.md) token standards, we reckon more can be done to aid the community in discovering alternative uses for NFTs.
+
+## Specification
+
+_The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119._
+
+```
+pragma solidity ^0.8.0;
+
+interface IEIP6066 {
+    /**
+     * @dev Should return whether the signature provided is valid for the provided tokenId and hash
+     * @param tokenId   Token ID of the signing NFT
+     * @param hash      Hash of the data to be signed
+     * @param data      Optional arbitrary data that may aid verification
+     *
+     * MUST return the bytes4 magic value 0xdff13226 when function passes.
+     * MUST NOT modify state (using STATICCALL for solc < 0.5, view modifier for solc > 0.5)
+     * MUST allow external calls
+     *
+     */
+    function isValidSignature(
+        uint256 tokenId,
+        bytes32 hash,
+        bytes calldata data
+    ) external view returns (bytes4 magicValue);
+}
+```
+
+`isValidSignature` can call arbitrary methods to validate a given signature.
+
+This function should be implemented by [EIP-721](./eip-721.md) or [EIP-1155](./eip-1155.md) compliant contracts that desire to enable its token holders to sign messages using their NFTs. Applications wanting to support contract signatures should call this method if the signer is an NFT.
+
```

Comment:
Optional ideas for backward compatibility with ERC1271

```suggestion
3. When a Compliant Contract also implements ERC1271 signature, the signing algorithm MUST be the identical signing algorithm with its choice of ERC1271.
```

Reply:
Mmm, I'm not so sure about this one. There isn't an explicit "signing algorithm" for this EIP, as consent is provided by the fact that the owner of the NFT created, signed, and broadcasted the transaction that would flip the boolean flag given a hash.

Reply:
Sure. I am open to not include this particular idea.